add SEUserByName, fallback to failsafe context in GetDefaultContextWithLevel

capnspacehook opened the pull request:
Ported getseuserbyname (https://github.com/SELinuxProject/selinux/blob/main/libselinux/src/seusers.c#L188) to Go, and added `failsafe_context` parsing to `GetDefaultContextWithLevel` as that's what libselinux's get_default_context_with_level does (https://github.com/SELinuxProject/selinux/blob/main/libselinux/src/get_context_list.c#L488).

`getseuserbyname` reads the`seusers` file (https://www.man7.org/linux/man-pages/man5/seusers.5.html) to find the SELinux user and the MLS level for a given Linux user.

`get_default_context_with_level` takes a SELinux user, MLS level, and an SELinux context (combination of SELinux user, role, domain, and MLS level, in many cases this context is the context of the caller) and returns the SELinux context that should be used when creating processes as a certain Linux user.

`get_default_context_with_level` attempts to find a suitable context for the specific SELinux user that was passed first, then searches in the list of global context mappings if no match was found, and finally just returns the failsafe context if no match was found anywhere else. `GetDefaultContextWithLevel` previously returned an error if no context could be found for the SELinux user or globally.

Man page for failsafe_context: https://www.man7.org/linux//man-pages/man5/failsafe_context.5.html

capnspacehook commented:
@rhatdan no prob, signed commit

kolyshkin commented:
@capnspacehook can you split this into two commits, by functionality. Same PR is fine.

capnspacehook committed:
add SEUserByName

Ported getseuserbyname (https://github.com/SELinuxProject/selinux/blob/main/libselinux/src/seusers.c#L188)
to Go, and added `failsafe_context` parsing to `GetDefaultContextWithLevel`
as that's what libselinux's get_default_context_with_level does
(https://github.com/SELinuxProject/selinux/blob/main/libselinux/src/get_context_list.c#L488).

`getseuserbyname` reads the`seusers` file (https://www.man7.org/linux/man-pages/man5/seusers.5.html)
to find the SELinux user and the MLS level for a given Linux user.

`get_default_context_with_level` takes a SELinux user, MLS level, and an
SELinux context (combination of SELinux user, role, domain, and MLS level,
in many cases this context is the context of the caller) and returns the
SELinux context that should be used when creating processes as a certain
Linux user.

`get_default_context_with_level` attempts to find a suitable context for
the specific SELinux user that was passed first, then searches in the list
of global context mappings if no match was found, and finally just returns
the failsafe context if no match was found anywhere else.
`GetDefaultContextWithLevel` previously returned an error if no context
could be found for the SELinux user or globally.

Man page for failsafe_context: https://www.man7.org/linux//man-pages/man5/failsafe_context.5.html

Signed-off-by: Andrew LeFevre <andrew.lefevre@goteleport.com>

kolyshkin reviewed (commented):
Regarding the second patch -- I see we're opening two files in `getDefaultContextWithLevel`, while we might not even read the second one. Your patch adds a third file to open.

Maybe we can rework the code first to open the file only when needed (the most changes will be in the tests I guess), then add the failsafe context?

capnspacehook commented:
@rhatdan done, do you need me to do anything else?

rhatdan commented:
Rebase and squash your commits, and we can merge.
